indra/llui/llluafloater.cpp

```diff
@@ -0,0 +1,271 @@
+/** 
+ * @file llluafloater.cpp
+ *
+ * $LicenseInfo:firstyear=2024&license=viewerlgpl$
+ * Second Life Viewer Source Code
+ * Copyright (C) 2024, Linden Research, Inc.
+ * 
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation;
+ * version 2.1 of the License only.
+ * 
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ * 
+ * Linden Research, Inc., 945 Battery Street, San Francisco, CA  94111  USA
+ * $/LicenseInfo$
+ *
+ */
+
+#include "llluafloater.h"
+
+#include <filesystem>
+#include "llevents.h"
+
+#include "llcheckboxctrl.h"
+#include "llcombobox.h"
+#include "llscrolllistctrl.h"
+
+const std::string LISTENER_NAME("LLLuaFloater");
+
+std::map<std::string, std::string> EVENT_LIST = 
+{
+    {"COMMIT_EVENT", "commit"},
+    {"DOUBLE_CLICK_EVENT", "double_click"},
+    {"MOUSE_ENTER_EVENT", "mouse_enter"},
+    {"MOUSE_LEAVE_EVENT", "mouse_leave"},
+    {"MOUSE_DOWN_EVENT", "mouse_down"},
+    {"MOUSE_UP_EVENT", "mouse_up"},
+    {"RIGHT_MOUSE_DOWN_EVENT", "right_mouse_down"},
+    {"RIGHT_MOUSE_UP_EVENT", "right_mouse_up"},
+    {"POST_BUILD_EVENT", "post_build"},
+    {"CLOSE_EVENT", "floater_close"}
+};
```

Comment:
At first, I used const std::string variables for all UI events, but then I realized that it will be good to be able to request events list from the script. So, it was turned into std::map. 
Should I turn it into the list and use just for informing the script about possible events? Maybe you can suggest something cleaner, because I agree that currently it's far from optimal.

Reply:
I was thinking a `std::vector<std::string>` would work. But I guess I see your point: you want to guarantee that an event referenced by code is definitely within `EVENT_LIST` -- eliminating the possibility of a typo in a string event name, or introducing a new correctly-spelled event name but forgetting to add it to `EVENT_LIST`.

There's a risk with this approach, though. `operator[]()` requires a non-`const` `map` -- but that means that a typo in the event _key_ name would add an entry to `EVENT_LIST` and send the empty string as the event name. You could make `EVENT_LIST` `const` and write `EVENT_LIST.find("CLOSE_EVENT")->second`, but that loses much of the cleanliness of `EVENT_LIST["CLOSE_EVENT"]` -- and with _that_ code, a typo in the key name would result in Undefined Behavior when you dereferenced `end()`.

How about a `post_event()` method that accepts the `LLSD` payload and also a `std::string_view` event name? `post_event()` would stamp `payload["event"] = event_name` -- but _first_ it would `assert()` that `event_name` is in `EVENT_LIST`. `EVENT_LIST` could be a `std::set<std::string>`.

Similarly you could have an `event_is(event, "mouse_down")` that verifies the event name is in `EVENT_LIST` before returning the `bool` result of the comparison.